File: robot_control/src/odom.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import rospy
import tf
import tf2_ros
import tf_conversions
from geometry_msgs.msg import TransformStamped
from geometry_msgs.msg import Vector3
from geometry_msgs.msg import Quaternion
from gazebo_msgs.srv import GetModelState
from nav_msgs.msg import Odometry
from std_msgs.msg import Header

def euler_to_quaternion(euler):
    q = tf.transformations.quaternion_from_euler(euler.x, euler.y, euler.z)
    return Quaternion(x=q[0], y=q[1], z=q[2], w=q[3])

def quaternion_to_euler(quaternion):
    e = tf.transformations.euler_from_quaternion((quaternion.x, quaternion.y, quaternion.z, quaternion.w))
    return Vector3(x=e[0], y=e[1], z=e[2])

if __name__ == '__main__':
    rospy.init_node('odom_broadcaster')
    pub = rospy.Publisher('/4omni/controller/odom',Odometry,queue_size=10)
    state_prox = rospy.ServiceProxy('/gazebo/get_model_state',GetModelState)
    while not rospy.is_shutdown():
        r = Odometry()
        r.header = Header()
        r.header.stamp = rospy.Time.now()
        r.header.frame_id = "4omni/odom"
        r.child_frame_id ="4omni/base_link"
        rospy.wait_for_service('/gazebo/get_model_state')
        state = state_prox('4omni','world')
        euler = quaternion_to_euler(state.pose.orientation)
        euler.z *= 0.9
        r.pose.pose.orientation = euler_to_quaternion(euler)
        r.pose.pose.position = state.pose.position
        r.pose.covariance = [0.001, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.001, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1000000.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1000000.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1000000.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1000.0]
        r.twist.twist = state.twist
        r.twist.covariance =[0.001, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.001, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1000000.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1000000.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1000000.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1000.0]
        pub.publish(r)
        rospy.loginfo('Transform Published')
        rospy.Rate(200.0).sleep()

File: robot_control/src/odom.cpp
#include "ros/ros.h"
#include "std_msgs/Float64.h"
#include "std_msgs/Float32.h"
#include "nav_msgs/Odometry.h"
#include "geometry_msgs/PoseStamped.h"

#include "tf/tf.h"
#include "tf/transform_broadcaster.h"

#include <string>

class OdomBroadcaster
{
private:
    ros::NodeHandle nh;

    ros::Subscriber odom_yaw_sub;
    ros::Subscriber odom_x_sub;
    ros::Subscriber odom_y_sub;

    nav_msgs::Odometry odom_msg;
    ros::Publisher odom_pub;

    ros::Timer control_tim;

    std::string odom_frame;
    std::string base_frame;

    double _yaw;
    double _x;
    double _y;

    void odomYawCallback(const std_msgs::Float32::ConstPtr& yaw);
    void odomXCallback(const std_msgs::Float32::ConstPtr& x);
    void odomYCallback(const std_msgs::Float32::ConstPtr& y);

    void tf_publish(geometry_msgs::Pose);

    void TimerCallback(const ros::TimerEvent& event);

public:
    OdomBroadcaster(void);
};

OdomBroadcaster::OdomBroadcaster(void)
{
    auto _nh = ros::NodeHandle("~");

    _yaw = _x = _y = 0.0;

    odom_yaw_sub = nh.subscribe<std_msgs::Float32>("odom/yaw", 10, &OdomBroadcaster::odomYawCallback, this);
    odom_x_sub = nh.subscribe<std_msgs::Float32>("odom/x", 10, &OdomBroadcaster::odomXCallback, this);
    odom_y_sub = nh.subscribe<std_msgs::Float32>("odom/y", 10, &OdomBroadcaster::odomYCallback, this);

    odom_pub = nh.advertise<nav_msgs::Odometry>("odom", 1);

    _nh.param<std::string>("odom_frame", odom_frame, "odom");
    _nh.param<std::string>("base_frame", base_frame, "base_link");

    int ctrl_freq;
    _nh.param("publish_rate", ctrl_freq, 20);

    control_tim = nh.createTimer(ros::Duration(1.0 / ctrl_freq), &OdomBroadcaster::TimerCallback, this);
}

void OdomBroadcaster::odomYawCallback(const std_msgs::Float32::ConstPtr& yaw)
{
    _yaw = (double)yaw->data;
}

void OdomBroadcaster::odomXCallback(const std_msgs::Float32::ConstPtr& x)
{
    _x = (double)x->data;
}

void OdomBroadcaster::odomYCallback(const std_msgs::Float32::ConstPtr& y)
{
    _y = (double)y->data;
}

void OdomBroadcaster::tf_publish(geometry_msgs::Pose pose0){
  static tf::TransformBroadcaster br;
  tf::Transform transform;
  tf::poseMsgToTF(pose0,transform);
  br.sendTransform(tf::StampedTransform(transform, ros::Time::now(), odom_frame, base_frame));
}

void OdomBroadcaster::TimerCallback(const ros::TimerEvent& event)
{

    odom_msg.pose.pose.orientation = tf::createQuaternionMsgFromYaw(_yaw);

    odom_msg.pose.pose.position.x = _x;
    odom_msg.pose.pose.position.y = _y;

    odom_msg.header.frame_id = odom_frame;
    odom_msg.child_frame_id= base_frame;

    odom_msg.header.stamp = ros::Time::now();

    odom_pub.publish(odom_msg);
    tf_publish(odom_msg.pose.pose);
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "odom");

    OdomBroadcaster *odomBroadcaster = new OdomBroadcaster();
    ROS_INFO("odom_broadcaster node has started.");

    ros::spin();
    ROS_INFO("odom_broadcaster node has been terminated.");
}
